<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [36]:
import pandas as pd
import psycopg2

In [1]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

In [38]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [39]:
query_3_1 = f'''
SELECT COUNT(*) AS vacancies_count
FROM vacancies
'''

In [40]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


,vacancies_count
0,49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [41]:
query_3_2 = f'''
SELECT COUNT(*) AS employers_count
FROM employers
'''

In [42]:
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\3588252056.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


,employers_count
0,23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [43]:
query_3_3 = f'''
SELECT COUNT(*) AS areas_count
FROM areas
'''

In [44]:
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\2018224393.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


,areas_count
0,1362


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [45]:
query_3_4 = f'''
SELECT COUNT(*) AS industries_count
FROM industries
'''

In [46]:
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\2172479966.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


,industries_count
0,294


***

*Общий вывод по предварительному анализу данных*

В базе представлено 49K вакансий, 23.5K работодателей, 1362 региона и 294 отрасли, что говорит о достаточно широком и диверсифицированном рынке труда.

 Соотношение вакансий к работодателям (~2:1) указывает на то, что многие компании размещают более одной вакансии, но при этом рынок не является сильно монополизированным.

Большое количество регионов при относительно меньшем числе вакансий означает сильную неравномерность распределения спроса по географии (ожидаем концентрацию в крупных городах).

Наличие 294 сфер деятельности подтверждает, что данные охватывают разные сегменты рынка, что важно для дальнейшего анализа специализаций

Вывод:
Рынок представлен широко, но изначально видно, что он фрагментирован по регионам и, вероятно, сконцентрирован в ключевых экономических центрах

Широкая география вакансий и разнообразие отраслей создают хорошие условия для дальнейшего анализа:

- распределения вакансий по регионам,

- структуры спроса на специалистов,

- требований работодателей,

- особенностей рынка труда в различных сферах экономики.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [47]:
query_4_1 = f'''
-- Подсчитываем количество вакансий в каждом регионе
SELECT
    a.name AS area_name,
    COUNT(v.id) AS cnt
FROM vacancies AS v
JOIN areas AS a ON v.area_id = a.id
GROUP BY a.name
ORDER BY cnt DESC
'''

In [48]:
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\347208320.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


,area_name,cnt
0,Москва,5333
1,Санкт-Петербург,2851
2,Минск,2112
3,Новосибирск,2006
4,Алматы,1892
...,...,...
764,Тарко-Сале,1
765,Новоаннинский,1
766,Бирск,1
767,Сасово,1


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [49]:
query_4_2 = f'''
-- Подсчитываем количество вакансий, в которых указана хотя бы одна граница зарплаты
SELECT
    COUNT(id) AS cnt
FROM vacancies
WHERE salary_from IS NOT NULL
   OR salary_to IS NOT NULL
'''

In [50]:
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\2168503709.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


,cnt
0,24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [51]:
query_4_3 = f'''
-- Рассчитываем средние значения нижней и верхней границ зарплаты
SELECT
    ROUND(AVG(salary_from), 0) AS avg_salary_from,
    ROUND(AVG(salary_to), 0) AS avg_salary_to
FROM vacancies
'''

In [52]:
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\2222372203.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


,avg_salary_from,avg_salary_to
0,71065.0,110537.0


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [53]:
query_4_4 = f'''
-- Группируем вакансии по графику работы и типу занятости
-- Считаем количество вакансий в каждой комбинации
SELECT
    schedule,
    employment,
    COUNT(id) AS cnt
FROM vacancies
GROUP BY
    schedule,
    employment
ORDER BY cnt DESC
'''

In [54]:
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\2981602351.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


,schedule,employment,cnt
0,Полный день,Полная занятость,35367
1,Удаленная работа,Полная занятость,7802
2,Гибкий график,Полная занятость,1593
3,Удаленная работа,Частичная занятость,1312
4,Сменный график,Полная занятость,940
5,Полный день,Стажировка,569
6,Вахтовый метод,Полная занятость,367
7,Полный день,Частичная занятость,347
8,Гибкий график,Частичная занятость,312
9,Полный день,Проектная работа,141


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [55]:
query_4_5 = f'''
-- Подсчитываем количество вакансий для каждого требуемого уровня опыта
SELECT
    experience,
    COUNT(id) AS cnt
FROM vacancies
GROUP BY experience
ORDER BY cnt ASC
'''

In [56]:
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\2601166264.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


,experience,cnt
0,Более 6 лет,1337
1,Нет опыта,7197
2,От 3 до 6 лет,14511
3,От 1 года до 3 лет,26152


***

Анализ вакансий показывает, что рынок труда характеризуется выраженной географической концентрацией. Наибольшее количество вакансий сосредоточено в крупных городах - прежде всего в Москве и Санкт-Петербурге, что объясняется высокой плотностью бизнеса, присутствием крупных компаний и развитой инфраструктурой. При этом, несмотря на широкое покрытие регионов, в большинстве из них наблюдается значительно меньшее количество вакансий, что подтверждает централизацию рынка.

Анализ зарплатных данных показывает, что средняя нижняя граница составляет около 71 тыс. рублей, а верхняя - около 110 тыс. рублей. При этом менее половины вакансий (~49%) содержат информацию о зарплате, что указывает на низкую прозрачность рынка и затрудняет объективную оценку уровня доходов. Широкий диапазон зарплат отражает различия в требованиях, уровне опыта и условиях найма.

Структура занятости демонстрирует явное доминирование полной занятости с графиком полного рабочего дня. Альтернативные форматы (удалённая работа, гибкий график, частичная занятость) представлены существенно реже, однако их наличие свидетельствует о постепенной трансформации рынка в сторону большей гибкости.

Таким образом, рынок вакансий можно охарактеризовать как централизованный, ориентированный на стандартную занятость и обладающий ограниченной прозрачностью по уровню заработной платы. Полученные результаты формируют основу для дальнейшего анализа, включая исследование зависимости зарплат от опыта, навыков и региональной специфики.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [57]:
query_5_1 = f'''
-- Определяем работодателей, занимающих 1-е и 5-е места
-- по количеству опубликованных вакансий
(
    SELECT
        e.name AS employer_name,
        COUNT(v.id) AS cnt
    FROM vacancies AS v
    JOIN employers AS e
        ON v.employer_id = e.id
    GROUP BY e.name
    ORDER BY cnt DESC
    LIMIT 1
)

UNION

(
    SELECT
        e.name AS employer_name,
        COUNT(v.id) AS cnt
    FROM vacancies AS v
    JOIN employers AS e
        ON v.employer_id = e.id
    GROUP BY e.name
    ORDER BY cnt DESC
    LIMIT 1 OFFSET 4
)
'''

In [58]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\1792814825.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


,employer_name,cnt
0,Газпром нефть,331
1,Яндекс,1933


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [59]:
query_5_2 = f'''
-- Для каждого региона считаем количество работодателей и вакансий
-- Оставляем только те регионы, где вакансий нет,
-- и выбираем регион с максимальным числом работодателей
SELECT
    a.name AS region_name,
    COUNT(DISTINCT e.id) AS cnt_employers,
    COUNT(DISTINCT v.id) AS cnt_vacancies
FROM areas AS a
LEFT JOIN employers AS e
    ON a.id = e.area
LEFT JOIN vacancies AS v
    ON a.id = v.area_id
GROUP BY
    a.id,
    a.name
HAVING COUNT(DISTINCT v.id) = 0
ORDER BY cnt_employers DESC
LIMIT 1
'''

In [60]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\3341481329.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


,region_name,cnt_employers,cnt_vacancies
0,Россия,410,0


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [61]:
query_5_3 = f'''
-- Подсчитываем, в скольких регионах каждый работодатель разместил вакансии
-- Используем DISTINCT, чтобы один и тот же регион не учитывался несколько раз
SELECT
    e.id,
    e.name,
    COUNT(DISTINCT v.area_id) AS areas_count
FROM employers AS e
LEFT JOIN vacancies AS v
    ON e.id = v.employer_id
GROUP BY
    e.id,
    e.name
ORDER BY areas_count DESC
'''

In [62]:
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\2398953733.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


,id,name,areas_count
0,1740,Яндекс,181
1,2748,Ростелеком,152
2,5724811,Спецремонт,116
3,5130287,Поляков Денис Иванович,88
4,3682876,ООО ЕФИН,71
...,...,...,...
23496,5287515,Но Смокинг,0
23497,2839354,МАСТЕРСКАЯ CREED,0
23498,5774578,Dzing Finance,0
23499,5775069,КОРС,0


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [63]:
query_5_4 = f'''
-- Подсчитываем количество работодателей,
-- для которых не указана сфера деятельности
SELECT
    COUNT(e.id) AS employers_without_industry
FROM employers AS e
LEFT JOIN employers_industries AS ei
    ON e.id = ei.employer_id
WHERE ei.industry_id IS NULL
'''

In [64]:
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\3114411492.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


,employers_without_industry
0,8419


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [65]:
query_5_5 = f'''
-- Находим компанию, которая стоит третьей по алфавиту
-- среди работодателей, у которых указано ровно 4 сферы деятельности
SELECT
    e.name
FROM employers AS e
JOIN employers_industries AS ei
    ON e.id = ei.employer_id
GROUP BY
    e.id,
    e.name
HAVING COUNT(ei.industry_id) = 4
ORDER BY e.name
OFFSET 2
LIMIT 1
'''

In [66]:
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\1650419958.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


,name
0,2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [67]:
query_5_6 = f'''
-- Подсчитываем количество работодателей,
-- относящихся к сфере "Разработка программного обеспечения"
SELECT
    COUNT(e.id) AS employers_count
FROM employers AS e
JOIN employers_industries AS ei
    ON e.id = ei.employer_id
JOIN industries AS i
    ON ei.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения'
'''

In [68]:
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\1169039133.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


,employers_count
0,3553


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [69]:
!pip install html5lib

Defaulting to user installation because normal site-packages is not writeable


In [70]:
import requests

# Wiki link
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'

# Добавляем заголовок User-Agent, чтобы запрос выглядел как запрос из браузера
headers = {
    'User-Agent': (
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
        'AppleWebKit/537.36 (KHTML, like Gecko) '
        'Chrome/114.0.0.0 Safari/537.36'
    )
}

# HTML страничка
response = requests.get(url, headers=headers)

# Считываем таблицы со страницы и выбираем таблицу,
# содержащую столбец с названием города
tables = pd.read_html(response.text, match='Город')

# Нужная таблица первая по счету
df_cities = tables[0]

# Извлекаем названия городов-миллионников в список
million_cities = df_cities['Город'].tolist()

# Удаляем сноски из названий городов
million_cities = [city.split('[')[0] for city in million_cities]

print(million_cities)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


C:\Users\user\AppData\Local\Temp\ipykernel_39572\967703430.py:20: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  tables = pd.read_html(response.text, match='Город')


In [71]:
cities_tuple = tuple(million_cities)

query_5_7 = f'''
-- Подсчитываем количество вакансий компании Яндекс
-- по городам-миллионникам, а также добавляем общий итог
(
    SELECT
        a.name AS region_name,
        COUNT(v.id) AS cnt
    FROM vacancies AS v
    JOIN areas AS a
        ON v.area_id = a.id
    JOIN employers AS e
        ON v.employer_id = e.id
    WHERE e.name = 'Яндекс'
      AND a.name IN {cities_tuple}
    GROUP BY a.name
)

UNION ALL

(
    SELECT
        'Total' AS region_name,
        COUNT(*) AS cnt
    FROM vacancies AS v
    JOIN areas AS a
        ON v.area_id = a.id
    JOIN employers AS e
        ON v.employer_id = e.id
    WHERE e.name = 'Яндекс'
      AND a.name IN {cities_tuple}
)

ORDER BY cnt ASC;
'''

In [72]:
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\3129374555.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


,region_name,cnt
0,Омск,21
1,Челябинск,22
2,Красноярск,23
3,Волгоград,24
4,Пермь,25
5,Казань,25
6,Ростов-на-Дону,25
7,Уфа,26
8,Самара,26
9,Краснодар,30


***

Анализ работодателей показывает, что рынок вакансий характеризуется высокой концентрацией спроса у крупных компаний и неравномерным распределением активности между участниками рынка.

Лидером по количеству размещённых вакансий является Яндекс (1933 вакансии), значительно опережающий других работодателей, что указывает на масштабную и постоянную потребность в кадрах, а также широкую географию найма. Пятое место занимает Газпром нефть, что подтверждает высокую активность крупных корпоративных игроков. Существенный разрыв между лидером и остальными компаниями свидетельствует о том, что именно крупнейшие работодатели формируют значительную долю рынка.

Регион «Россия» выступает как агрегированное или техническое значение без привязки к конкретному месту публикации вакансий, поэтому отсутствие вакансий в нём связано с особенностями структуры данных, а не с реальным отсутствием спроса.

Анализ географии публикаций демонстрирует, что крупнейшие компании обладают широкой региональной представленностью. Так, Яндекс размещает вакансии в 181 регионе, что подтверждает масштаб бизнеса и распределённую модель найма. В то же время значительная часть работодателей либо ограничена одним регионом, либо не ведёт активного найма, что отражает локальный характер бизнеса или нерегулярность публикаций.

С точки зрения структуры данных выявлена существенная доля неполных записей: у 8419 работодателей не указана сфера деятельности. Это снижает точность отраслевого анализа и требует осторожности при интерпретации результатов. При этом 3553 работодателя относятся к сфере разработки программного обеспечения, что подчёркивает значительную долю IT-сектора и высокий спрос на технических специалистов.

Дополнительный анализ вакансий Яндекса в городах-миллионниках показывает, что основная активность сосредоточена в Москве и Санкт-Петербурге, а суммарно в крупнейших городах размещено 485 вакансий, что подтверждает общую тенденцию концентрации спроса в ключевых экономических и технологических центрах.

Итог:
Рынок работодателей характеризуется доминирующей ролью крупных компаний, широкой географической экспансией лидеров, высокой долей IT-сектора и наличием ограничений в качестве данных. Основной спрос формируется крупными технологическими и корпоративными работодателями, при этом активность локальных компаний значительно ниже и менее масштабна.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [73]:
query_6_1 = f'''
-- Подсчитываем количество вакансий, в названии которых встречаются
-- упоминания data или "данн" 
SELECT
    COUNT(*) AS cnt
FROM vacancies
WHERE name ILIKE '%data%'
   OR name ILIKE '%данн%';
'''

#WHERE name ~* 'data|данн';
#WHERE LOWER(name) LIKE '%data%'
   #OR LOWER(name) LIKE '%данн%';

In [74]:
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\485020324.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


,cnt
0,1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [75]:
query_6_2 = f'''
-- Подсчитываем количество вакансий для начинающего Data Scientist-а 

SELECT COUNT(*)
FROM vacancies
WHERE
    (
        name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%'
        OR (
            name ILIKE '%ml%'
            AND name NOT ILIKE '%html%'
        )
    )
    AND
    (
        name ILIKE '%junior%'
        OR experience = 'Нет опыта'
        OR employment = 'Стажировка'
    )
    '''
    

In [76]:
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\3641650667.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


,count
0,51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [77]:
query_6_3 = f'''
-- Подсчитываем количество вакансий для Data Scientist-а,
-- где среди ключевых навыков указаны SQL или Postgres
SELECT COUNT(*)
FROM vacancies
WHERE
    (
        name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%'
        OR (
            name ILIKE '%ml%'
            AND name NOT ILIKE '%html%'
        )
    )
    AND
    (
        key_skills ILIKE '%sql%'
        OR key_skills ILIKE '%postgres%'
    )
    '''

In [78]:
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\1159587760.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


,count
0,229


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [79]:
query_6_4 = f'''
-- Подсчитываем количество вакансий для Data Scientist-а,
-- где среди ключевых навыков упоминается Python

SELECT COUNT(*)
FROM vacancies
WHERE
    (
        name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%'
        OR (
            name ILIKE '%ml%'
            AND name NOT ILIKE '%html%'
        )
    )
    AND
    (
        key_skills ILIKE '%python%'
        OR key_skills ILIKE '%пайтон%'
    )
    '''


In [80]:
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\972386973.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


,count
0,357


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [81]:
query_6_5 = f'''
-- Рассчитываем среднее количество ключевых навыков в вакансиях для Data Scientist-а

SELECT 
    ROUND(AVG(skill_count), 2) AS avg_skills
FROM (
    SELECT
        LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1 AS skill_count
    FROM vacancies
    WHERE
        key_skills IS NOT NULL
        AND key_skills != ''
        AND (
            name ILIKE '%data scientist%'
            OR name ILIKE '%data science%'
            OR name ILIKE '%исследователь данных%'
            OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
            OR name ILIKE '%machine learning%'
            OR name ILIKE '%машинн%обучен%'
        )
) AS ds_vacancies

''' 

In [82]:
df = pd.read_sql_query(query_6_5, connection) 
df


C:\Users\user\AppData\Local\Temp\ipykernel_39572\663933719.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


,avg_skills
0,6.41


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [83]:
query_6_6 = f'''
-- Рассчитываем среднюю зарплату для вакансий Data Scientist
-- с опытом работы от 3 до 6 лет
-- Если указаны обе границы зарплаты, берём их среднее значение;
-- если указана только одна, COALESCE подставляет её вместо пропущенной
SELECT 
    ROUND(AVG(
        (COALESCE(salary_from, salary_to) + 
         COALESCE(salary_to, salary_from)) / 2
    )) AS avg_salary
FROM vacancies
WHERE
    (
        name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%'
    )
    AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
    AND experience = 'От 3 до 6 лет';

'''

In [84]:
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_39572\3561045109.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


,avg_salary
0,243115.0


***

Проведённый анализ показывает, что в базе представлено 1771 вакансия, связанная с данными, что подтверждает устойчивый спрос на специалистов в области Data Science и смежных направлений. При этом только 51 вакансия относится к начальному уровню, что указывает на явный дисбаланс: рынок ориентирован преимущественно на специалистов с опытом, а порог входа в профессию остаётся достаточно высоким.

Анализ требований к навыкам подтверждает ключевую роль базовых инструментов Data Science. Навыки работы с базами данных (SQL/PostgreSQL) требуются в 229 вакансиях, а Python — в 357 вакансиях, что делает его наиболее востребованным инструментом. Это отражает стандарт индустрии: Python используется как основной язык для анализа данных и машинного обучения, а SQL - как необходимая база для работы с данными.

В среднем работодатели указывают 6.41 ключевого навыка на вакансию, что говорит о высокой комплексности требований. Фактически ожидается T-shaped специалист: не узкая специализация, а сочетание программирования, аналитики, работы с данными и ML-инструментами.

Анализ заработных плат показывает, что специалисты уровня middle (3–6 лет опыта) могут рассчитывать в среднем на ~243 тыс. рублей, что подтверждает высокую ценность таких специалистов. Это также объясняет низкую долю junior-вакансий: работодатели стремятся минимизировать затраты на обучение и сразу нанимать готовых специалистов.

# Общий вывод по проекту

Проведённый анализ вакансий показал, что рынок специалистов по работе с данными характеризуется устойчивым спросом и высокой значимостью аналитических компетенций для бизнеса. В базе данных выявлено значительное количество вакансий, связанных с анализом данных, что подтверждает активное развитие data-driven подхода в компаниях различных отраслей.

Анализ требований работодателей показывает, что профессия Data Scientist требует широкого набора компетенций: в среднем в вакансиях указывается более шести ключевых навыков. Наиболее востребованными инструментами являются Python и SQL, что отражает их роль как базового технологического стека для работы с данными - от извлечения и обработки до анализа и построения моделей.

Структура вакансий указывает на относительно высокий порог входа в профессию: доля предложений для начинающих специалистов существенно ниже, чем для кандидатов с опытом. Это связано с высокой сложностью задач и ожиданием от кандидатов не только теоретических знаний, но и практического опыта работы с реальными данными и инструментами.

Анализ заработных плат подтверждает высокую ценность специалистов Data Science для бизнеса. Для специалистов уровня middle средняя предлагаемая заработная плата составляет около 243 тыс. рублей, что соответствует верхнему сегменту доходов в IT и отражает высокий спрос на квалифицированные кадры.

В целом результаты исследования показывают, что рынок Data Science характеризуется устойчивым спросом, высокими требованиями к навыкам и уровнем оплаты, опережающим многие другие направления. При этом успешный вход в профессию требует системной подготовки, практического опыта и владения ключевыми инструментами анализа данных.

Перспективным направлением дальнейшего исследования является анализ влияния набора навыков, отрасли работодателя и региональной специфики на уровень заработной платы, что позволит глубже понять факторы формирования спроса и стоимости специалистов.

# Юнит 7. Дополнительные исследования

In [85]:
import plotly.express as px

In [86]:
# Фильтр вакансий Data Scientist по ключевым словам в названии.
# Исключаем HTML-вакансии при поиске по сокращению ML.
ds_filter = '''
(
    v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%'
    OR (
        v.name ILIKE '%ml%'
        AND v.name NOT ILIKE '%html%'
    )
)
'''


# Выражение для расчета средней зарплаты:
# если указаны обе границы, берём среднее;
# если одна из границ отсутствует, COALESCE подставляет имеющееся значение.
salary_expr = '''
ROUND(AVG(
    (COALESCE(v.salary_from, v.salary_to) + COALESCE(v.salary_to, v.salary_from)) / 2.0
), 0)
'''


# Выражение для подсчета числа навыков в вакансии.
# Навыки в поле key_skills разделены табуляцией, поэтому
# количество навыков = количество разделителей + 1
skill_count_expr = '''
LENGTH(v.key_skills) - LENGTH(REPLACE(v.key_skills, CHR(9), '')) + 1
'''

In [87]:
query_7_1 = f'''
-- Исследуем различия в средней зарплате Data Scientist
-- по регионам и требуемому опыту работы
-- Учитываем только вакансии с указанной зарплатой
-- и оставляем группы, где есть не менее 3 наблюдений

SELECT
    a.name AS area_name,
    v.experience,
    COUNT(v.id) AS vacancies_count,
    {salary_expr} AS avg_salary
FROM vacancies AS v
JOIN areas AS a
    ON v.area_id = a.id
WHERE
    {ds_filter}
    AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)
GROUP BY
    a.name,
    v.experience
HAVING COUNT(v.id) >= 3
ORDER BY
    avg_salary DESC,
    vacancies_count DESC
'''

In [88]:
df_7_1 = pd.read_sql_query(query_7_1, connection)
df_7_1

C:\Users\user\AppData\Local\Temp\ipykernel_39572\3921586033.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_1 = pd.read_sql_query(query_7_1, connection)


,area_name,experience,vacancies_count,avg_salary
0,Москва,От 3 до 6 лет,13,274231.0
1,Санкт-Петербург,От 3 до 6 лет,3,271667.0
2,Армения,От 3 до 6 лет,3,268863.0
3,Санкт-Петербург,От 1 года до 3 лет,3,190115.0
4,Москва,От 1 года до 3 лет,14,169179.0
5,Санкт-Петербург,Нет опыта,3,87667.0


In [89]:
# Выбираем регионы с наибольшим суммарным количеством DS-вакансий,
# чтобы визуализация была компактной и содержательной
top_areas_7_1 = (
    df_7_1.groupby('area_name', as_index=False)['vacancies_count']
    .sum()
    .sort_values('vacancies_count', ascending=False)
    .head(15)
)

top_area_names_7_1 = top_areas_7_1['area_name'].tolist()

df_7_1_top = df_7_1[df_7_1['area_name'].isin(top_area_names_7_1)]

# Строим сводную таблицу для тепловой карты:
# строки - регионы, столбцы - опыт, значения - средняя зарплата.
pivot_7_1 = df_7_1_top.pivot(
    index='area_name',
    columns='experience',
    values='avg_salary'
)

fig = px.imshow(
    pivot_7_1,
    text_auto=True,
    aspect='auto',
    title='Средняя зарплата Data Scientist по регионам и опыту'
)

fig.update_layout(
    xaxis_title='Опыт работы',
    yaxis_title='Регион'
)

fig.show()

fig.write_html('pivot_7_1.html')

Анализ показывает выраженную зависимость уровня заработной платы от опыта работы: при переходе к категории специалистов с опытом 3–6 лет средняя зарплата существенно возрастает. Наиболее высокие значения наблюдаются в крупнейших технологических центрах - Москве и Санкт-Петербурге, что отражает концентрацию высокооплачиваемых позиций Data Science в крупных IT-хабах.

Интересно отметить, что в выборке присутствуют вакансии из других стран, например из Армении, что указывает на более широкую географию данных и свидетельствует о том, что рынок специалистов по анализу данных в рассматриваемой базе не ограничивается исключительно российскими регионами.

In [90]:
query_7_2 = f'''
-- Исследуем связь между количеством навыков,
-- требуемым опытом и средней зарплатой Data Scientist
-- Учитываем только вакансии с непустым списком навыков
-- и указанной зарплатой

SELECT
    v.experience,
    {skill_count_expr} AS skills_count,
    COUNT(v.id) AS vacancies_count,
    {salary_expr} AS avg_salary
FROM vacancies AS v
WHERE
    v.key_skills IS NOT NULL
    AND v.key_skills != ''
    AND {ds_filter}
    AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)
GROUP BY
    v.experience,
    {skill_count_expr}
HAVING COUNT(v.id) >= 3
ORDER BY
    v.experience,
    skills_count
'''

In [91]:
df_7_2 = pd.read_sql_query(query_7_2, connection)
df_7_2

C:\Users\user\AppData\Local\Temp\ipykernel_39572\382515501.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



,experience,skills_count,vacancies_count,avg_salary
0,От 1 года до 3 лет,5,7,122857.0
1,От 1 года до 3 лет,6,5,160000.0
2,От 1 года до 3 лет,8,4,152500.0
3,От 3 до 6 лет,3,7,237641.0
4,От 3 до 6 лет,4,4,312500.0
5,От 3 до 6 лет,5,11,280658.0
6,От 3 до 6 лет,6,3,212081.0
7,От 3 до 6 лет,8,4,215000.0
8,От 3 до 6 лет,9,4,154282.0


In [92]:
fig = px.scatter(
    df_7_2,
    x='skills_count',
    y='avg_salary',
    color='experience',
    size='vacancies_count',
    hover_data=['vacancies_count'],
    title='Зависимость зарплаты от количества навыков и опыта'
)

fig.update_layout(
    xaxis_title='Количество навыков',
    yaxis_title='Средняя зарплата'
)

fig.show()

fig.write_html('plots/pivot_7_2.html')

Результаты показывают, что уровень заработной платы в первую очередь определяется опытом работы, тогда как количество указанных навыков оказывает менее выраженное влияние. Для специалистов с опытом 3-6 лет зарплаты значительно выше независимо от числа навыков. Это указывает на то, что работодатели ценят прежде всего практический опыт и глубину владения ключевыми инструментами.

Для повышения надежности анализа учитываются только навыки, встречающиеся не менее чем в 3 вакансиях, чтобы исключить влияние случайных значений. (Задание 7.3 и 7.4)

In [93]:
query_7_3 = f'''
-- Сравниваем востребованность и среднюю зарплату
-- для отдельных ключевых навыков в DS-вакансиях
-- Учитываем только навыки, встречающиеся не менее чем в 3 вакансиях

SELECT *
FROM (
    SELECT
        'Python' AS skill,
        COUNT(v.id) AS vacancies_count,
        {salary_expr} AS avg_salary
    FROM vacancies AS v
    WHERE
        v.key_skills ILIKE '%Python%'
        AND {ds_filter}
        AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)

    UNION ALL

    SELECT
        'SQL' AS skill,
        COUNT(v.id) AS vacancies_count,
        {salary_expr} AS avg_salary
    FROM vacancies AS v
    WHERE
        v.key_skills ILIKE '%SQL%'
        AND {ds_filter}
        AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)

    UNION ALL

    SELECT
        'PostgreSQL' AS skill,
        COUNT(v.id) AS vacancies_count,
        {salary_expr} AS avg_salary
    FROM vacancies AS v
    WHERE
        v.key_skills ILIKE '%PostgreSQL%'
        AND {ds_filter}
        AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)

    UNION ALL

    SELECT
        'Machine Learning' AS skill,
        COUNT(v.id) AS vacancies_count,
        {salary_expr} AS avg_salary
    FROM vacancies AS v
    WHERE
        v.key_skills ILIKE '%Machine Learning%'
        AND {ds_filter}
        AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)

    UNION ALL

    SELECT
        'Pandas' AS skill,
        COUNT(v.id) AS vacancies_count,
        {salary_expr} AS avg_salary
    FROM vacancies AS v
    WHERE
        v.key_skills ILIKE '%Pandas%'
        AND {ds_filter}
        AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)
) AS skills_data

WHERE vacancies_count >= 3
ORDER BY avg_salary DESC
'''

In [94]:
df_7_3 = pd.read_sql_query(query_7_3, connection)
df_7_3

C:\Users\user\AppData\Local\Temp\ipykernel_39572\3099143861.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



,skill,vacancies_count,avg_salary
0,PostgreSQL,7,225962.0
1,Machine Learning,23,189803.0
2,Python,50,171922.0
3,SQL,37,169900.0
4,Pandas,9,152736.0


In [95]:
fig = px.bar(
    df_7_3.sort_values('avg_salary', ascending=True),
    x='avg_salary',
    y='skill',
    orientation='h',
    text='vacancies_count',
    title='Средняя зарплата по конкретным навыкам в вакансиях Data Scientist'
)

fig.update_layout(
    xaxis_title='Средняя зарплата',
    yaxis_title='Навык'
)

fig.show()

fig.write_html('plots/pivot_7_3.html')

Сравнение вакансий по отдельным навыкам показывает, что наиболее высокая средняя зарплата связана с технологиями работы с базами данных, в частности PostgreSQL. При этом Python и SQL остаются наиболее распространёнными навыками в вакансиях Data Scientist, выступая базовыми инструментами профессии. Таким образом, более специализированные технологии могут быть связаны с более высокой оплатой, но встречаются реже.

In [96]:
query_7_4 = f'''
-- Сравниваем комбинации навыков по числу вакансий и средней зарплате
-- Учитываем только комбинации, встречающиеся не менее чем в 3 вакансиях

SELECT *
FROM (
    SELECT
        'Python + SQL' AS skill_combination,
        COUNT(v.id) AS vacancies_count,
        {salary_expr} AS avg_salary
    FROM vacancies AS v
    WHERE
        v.key_skills ILIKE '%Python%'
        AND v.key_skills ILIKE '%SQL%'
        AND {ds_filter}
        AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)

    UNION ALL

    SELECT
        'Python + Pandas' AS skill_combination,
        COUNT(v.id) AS vacancies_count,
        {salary_expr} AS avg_salary
    FROM vacancies AS v
    WHERE
        v.key_skills ILIKE '%Python%'
        AND v.key_skills ILIKE '%Pandas%'
        AND {ds_filter}
        AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)

    UNION ALL

    SELECT
        'Python + Machine Learning' AS skill_combination,
        COUNT(v.id) AS vacancies_count,
        {salary_expr} AS avg_salary
    FROM vacancies AS v
    WHERE
        v.key_skills ILIKE '%Python%'
        AND v.key_skills ILIKE '%Machine Learning%'
        AND {ds_filter}
        AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)

    UNION ALL

    SELECT
        'SQL + PostgreSQL' AS skill_combination,
        COUNT(v.id) AS vacancies_count,
        {salary_expr} AS avg_salary
    FROM vacancies AS v
    WHERE
        v.key_skills ILIKE '%SQL%'
        AND v.key_skills ILIKE '%PostgreSQL%'
        AND {ds_filter}
        AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)

    UNION ALL

    SELECT
        'Python + PostgreSQL' AS skill_combination,
        COUNT(v.id) AS vacancies_count,
        {salary_expr} AS avg_salary
    FROM vacancies AS v
    WHERE
        v.key_skills ILIKE '%Python%'
        AND v.key_skills ILIKE '%PostgreSQL%'
        AND {ds_filter}
        AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)
) AS combinations_data

WHERE vacancies_count >= 3
ORDER BY vacancies_count DESC
'''

In [97]:
df_7_4 = pd.read_sql_query(query_7_4, connection)
df_7_4

C:\Users\user\AppData\Local\Temp\ipykernel_39572\2995428897.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



,skill_combination,vacancies_count,avg_salary
0,Python + SQL,25,154980.0
1,Python + Machine Learning,19,169163.0
2,Python + Pandas,9,152736.0
3,SQL + PostgreSQL,7,225962.0


In [98]:
fig = px.bar(
    df_7_4.sort_values('vacancies_count', ascending=True),
    x='vacancies_count',
    y='skill_combination',
    orientation='h',
    text='avg_salary',
    title='Частота и зарплата по комбинациям навыков Data Scientist'
)

fig.update_layout(
    xaxis_title='Количество вакансий',
    yaxis_title='Комбинация навыков'
)

fig.show()

fig.write_html('plots/pivot_7_4.html')

Наиболее распространённой комбинацией навыков в вакансиях Data Scientist является связка Python и SQL, что отражает фундаментальную роль этих инструментов в работе с данными. При этом более высокие уровни средней зарплаты наблюдаются для комбинаций навыков, связанных с инфраструктурой хранения данных, например SQL и PostgreSQL. Это может свидетельствовать о повышенной ценности специалистов, работающих на стыке анализа данных и инженерии данных.

In [99]:
query_7_5 = f'''
-- Определяем индустрии, в которых чаще всего встречаются
-- вакансии для Data Scientist

SELECT
    i.name AS industry_name,
    COUNT(DISTINCT v.id) AS vacancies_count
FROM vacancies AS v
JOIN employers AS e
    ON v.employer_id = e.id
JOIN employers_industries AS ei
    ON e.id = ei.employer_id
JOIN industries AS i
    ON ei.industry_id = i.id
WHERE
    {ds_filter}
GROUP BY
    i.name
ORDER BY
    vacancies_count DESC
LIMIT 15
'''

In [100]:
df_7_5 = pd.read_sql_query(query_7_5, connection)
df_7_5

C:\Users\user\AppData\Local\Temp\ipykernel_39572\402166413.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



,industry_name,vacancies_count
0,Разработка программного обеспечения,166
1,"Системная интеграция, автоматизации технологи...",156
2,Банк,96
3,"Интернет-компания (поисковики, платежные систе...",84
4,Мобильная связь,24
5,"Маркетинговые, рекламные, BTL, дизайнерские, E...",23
6,"Управляющая, инвестиционная компания (управлен...",21
7,Консалтинговые услуги,17
8,Фиксированная связь,13
9,"Научно-исследовательская, научная, академическ...",9


In [101]:
df_7_5_top = df_7_5.head(10).copy()

# Длинные названия индустрий упрощаем для удобства визуализации
df_7_5_top['industry_short'] = [
    'Разработка ПО',
    'Автоматизация / IT-консалтинг',
    'Банк',
    'Медиа / интернет-ресурсы',
    'Мобильная связь',
    'Маркетинг / реклама / PR',
    'Управляющая компания',
    'Консалтинг',
    'Фиксированная связь',
    'Наука / исследования'
]

fig = px.bar(
    df_7_5_top.sort_values('vacancies_count', ascending=True),
    x='vacancies_count',
    y='industry_short',
    orientation='h',
    text='vacancies_count',
    title='Топ-10 индустрий с наибольшим спросом на Data Scientist'
)

fig.update_layout(
    height=550,
    xaxis_title='Количество вакансий',
    yaxis_title='Индустрия'
)

fig.show()

df_7_5_top[['industry_short', 'industry_name', 'vacancies_count']]

fig.write_html('plots/pivot_7_5.html')

Наибольший спрос на специалистов Data Science формируется в компаниях, занимающихся разработкой ПО и технологическими решениями. Значительная доля вакансий также приходится на банковский сектор и интернет-компании, что отражает активное использование аналитики данных для финансовых сервисов, персонализации и оптимизации бизнес-процессов. Это подтверждает ключевую роль Data Science в цифровых и технологических отраслях экономики.